byte_extract lowering of unions [blocks: #2068]

tautschnig opened the pull request:
We previously handled unions like PODs.

Only the last commit is new, the rest is borrowed from #4124.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
byte_extract lowering of unions

We previously handled unions like PODs.

tautschnig committed:
byte_extract lowering of arrays: optimise for constant index

No need to construct index_exprt when we can just pick the operand out of a
vector.

tautschnig committed:
byte_extract lowering of vectors and array cleanup

Refactor the code used for arrays to make it re-usable for vectors and arrays.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 19b8b80373145ac0c2e5454a6b691182af21b55e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100912495

tautschnig committed:
byte_extract lowering: also use a lower bound

We already limited extraction at an uppper bound, but when we know the offset we
can also start as late as possible. For structs, for example, we now do not
expand components that will be irrelevant to the final result.

tautschnig committed:
byte-operator lowering: rename element_width to {element,component}_bits

This makes the unit of measurement explicit and thus should help avoid
bits/bytes confusion.

tautschnig committed:
byte_extract lowering: Use optionalt<mp_integer> instead of nil_exprt

This makes the interface much more explicit about what is or isn't supported at
moment.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 761ad8a2cbd0d090c31496878bb1deb78d3047a5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100979304
